lua/entities/gmod_wire_expression2/core/custom/collcallback.lua

```diff
@@ -0,0 +1,190 @@
+	if registered_ents[ent] == nil then
+		registered_ents[ent] = {}
+		if ent.PhysicsCollide then
+			ent.PhysicsCollide = function(sel,data,collider)
+				entitiesCollide(ent,data)
+			end
+		else
+			ent:AddCallback("PhysicsCollide",entitiesCollide)
+		end
+	end
+	if not table.HasValue(registered_ents[ent],gate) then
+		table.insert(registered_ents[ent],gate)
+		gate.colProps = gate.colProps+1
+	else return end
+end
+
+function removeCallback(ent,gate)
+	if not registered_ents[ent] then return end
+	local tbl = registered_ents[ent]
+	for i = 1,#tbl,1 do
+		if tbl[i] == gate then
+			table.remove(registered_ents[ent],i)
+			gate.colProps = gate.colProps-1
+			return
+		end
+	end
+end
+
+__e2setcost(2)
+e2function number isCollClk()
+	if not self.colData then return 0 else return 1 end
+	
+end
+
+e2function number collIgnoreConstrained(constrained)
+	self.collIgnoreConstrained = constrained ~= 0
+end
+
+e2function number collFilter(filter)
+	self.collFilter = filter ~= 0
+end
+
+e2function number entity:hasCollDetection()
+	if not registered_ents[this] then return 0 else return 1 end
+	
+end
+
+e2function void collEnableEarlyClk(enable)
+	self.physicsEarlyCallback = enable ~= 0
+end
+
+__e2setcost(5)
+e2function entity collEntity()
+	if not self.colData then return NULL end
+	return self.colData.ourEnt
+end
+
+e2function entity collHitEntity()
+	if not self.colData then return NULL end
+	return self.colData.HitEntity
+end
+
+e2function number collDelta()
+	if not self.colData then return 0 end
+	return self.colData.DeltaTime
+end
+
+e2function number collSpeed()
+	if not self.colData then return 0 end
+	return self.colData.Speed
+end
+
+e2function vector collVelocity()
+	if not self.colData then return Vector(0,0,0) end
+	return self.colData.OurOldVelocity
+end
+
+e2function vector collHitVelocity()
+	if not self.colData then return Vector(0,0,0) end
+	return self.colData.TheirOldVelocity
+end
+
+e2function vector collPos()
+	if not self.colData then return Vector(0,0,0) end
+	return self.colData.HitPos
+end
+
+e2function vector collNormal()
+	if not self.colData then return Vector(0,0,0) end
+	return self.colData.HitNormal
+end
+
+e2function vector collEntityPos()
+	if not self.colData then return Vector(0,0,0) end
+	return self.colData.posFirst
+end
+
+e2function vector collHitEntityPos()
+	if not self.colData then return Vector(0,0,0) end
+	return self.colData.posSecond
+end
+
+e2function angle collEntityAng()
+	if not self.colData then return Angle(0,0,0) end
+	return self.colData.angleFirst
+end
+
+e2function angle collHitEntityAng()
+	if not self.colData then return Angle(0,0,0) end
+	return self.colData.angleSecond
+end
+
+__e2setcost(30)
+e2function void entity:addCollDetection()
+	addCallback(this,self)
+end
+
+e2function void entity:addCollDetection(entity gate)
+	if IsValid(gate) and gate:GetClass() == "gmod_wire_expression2" and isOwner(self,gate.context) then
```

Comment:
This can be shortcut:
````Lua
if not IsValid(gate) then return end
if gate:GetClass() ~= "gmod_wire_expression2" then return end
if not isOwner(self,gate.context) then return end
````